File: src/_03_BehavioralPattern/_03_11_Visitor/Pad.js
import Device from "./Device.js";

import Rectangle from "./Rectangle.js";
import Triangle from "./Triangle.js";
import Circle from "./Circle.js";
import Star from "./Star.js";

/**
 * @typedef { import("./Shape").default } Shape
 */

export default class Pad extends Device {
  constructor() {
    super();
  }

  /**
   * @override
   * @param { Shape } shape 
   */
  print(shape) {
    switch(shape.constructor) {
      case Rectangle: {
        console.log("* Print Rectangle To Pad");
        break;
      }

      case Triangle: {
        console.log("* Print Triangle To Pad");
        break;
      }

      case Circle: {
        console.log("* Print Circle To Pad");
        break;
      }

      case Star: {
        console.log("* Print Star To Pad");
        break;
      }
    }
  }
}

File: src/_03_BehavioralPattern/_03_11_Visitor/java/after/Pad.java
package _03_BehavioralPattern._03_11_Visitor.java.after;

public class Pad implements Device {
  
  @Override
  public void print(Rectangle rectangle) {
    System.out.println("& Print Rectangle To Pad");
  }
  
  @Override
  public void print(Circle circle) {
    System.out.println("& Print Circle To Pad");
  }

  @Override
  public void print(Triangle triangle) {
    System.out.println("& Print Triangle To Pad");
  }

  @Override
  public void print(Star start) {
    System.out.println("& Print Star To Pad");
  }
  
}
